src/language/dafnyInstallation.ts

```diff
@@ -89,15 +89,63 @@ export async function getCompilerRuntimePath(context: ExtensionContext): Promise
   return configuredPath;
 }
 
+// We cache the language server runtime path so that we don't need to copy it every time.
+let LanguageServerRuntimePath: string | null = null;
+
+function isNullOrEmpty(s: string | null): boolean {
+  return s === '' || s == null;
+}
+
 export async function getLanguageServerRuntimePath(context: ExtensionContext): Promise<string> {
```

Comment:
I want to point out here that you accidentally introduced a race condition here due to the asynchronous implementation.
It's possible that `LanguageServerRuntimePath` changes at any `await`.

Moreover: I suggest separating the `getLanguageServerRuntimePath` functionallity from the copying functionality. This function has 2 primary uses: Checking/retrieving the current Dafny installation when starting up as well as identifying the target directory for the automatic installation. The latter would fail but it looks like you've already respected this case with `mightBeRecomputed`. Maybe naming this variable to something like `isCustomInstallation` is more informative here.
Personally, I don't like `get*` functions that have such side-effects as most people wouldn't expect it, especially due to the name. I'm sure you can come up with a good idea to separate it since the only location where this is of any concern is in `dafnyLanguageClient.ts`.

Reply:
> It's possible that LanguageServerRuntimePath changes at any await.

No it's not in this case, because JS is single-threaded, and all the calls to getLanguageServerRuntimePath are within the same function call.

I renamed `getLanguageServerRuntimePath` to `getOrComputeLanguageServerRuntimePath` to indicate potential side-effects, and I would advise against anything that looks like we have to explicitly manage at which point we do the duplication of the dlls, and at which point we only retrieve the computed value. It would force adding parameters where we don't want to, or come up with class fields that can be nullable and then we have to deal with various warnings.
`getLanguageServerRuntimePath` is encapsulating the behavior, and it's not modifying anything a user of this function can access, so it's still a true get method.